### PR TITLE
[CIS-245] Event observing

### PR DESCRIPTION
**Pull request by evsaev**
**This PR:**

- introduces `MemberEvent`
- introduces `EventObserver` and `MemberEventObserver` for event-listening
- updates member events to be non-generic and have only generic `init`
- updates `ChannelController` to listen for and deliver `MemberEvent` from it's channel

### CODE REVIEW

File: StreamChatClient_v3/Workers/EventObservers/EventObserver.swift [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+    init<T>(notificationCenter: NotificationCenter, tranform: @escaping (Event) -> T?, callback: @escaping (T) -> Void) {
+        let observer = notificationCenter.addObserver(forName: .NewEventReceived, object: nil, queue: nil) {
+            guard let event = $0.event.flatMap(tranform) else { return }
+            callback(event)
+        }
+
+        stopObserving = { [weak notificationCenter] in
+            notificationCenter?.removeObserver(observer)
+        }
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Workers/EventObservers/EventObserver.swift:18]: That's a very nice trick!

File: StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    var eventToDeliver: HealthCheckEvent!
+    var eventNotification: Notification {
+        .init(newEventReceived: eventToDeliver, sender: self)
+    }
+
+    // MARK: - Setup
+
+    override func setUp() {
+        notificationCenter = NotificationCenter()
+        eventToDeliver = HealthCheckEvent(connectionId: .unique)
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift:19]: please also call `super` in `setUp` and `tearDown`

File: StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        // Create observer and count callback executions
+        var observer: EventObserver? = .init(notificationCenter: notificationCenter,
+                                             tranform: { $0 as? HealthCheckEvent },
+                                             callback: { _ in callbackExecutionCount += 1 })
+        _ = observer
+
+        // Send event notification
+        notificationCenter.post(eventNotification)
+        // Assert callback is called
+        XCTAssertEqual(callbackExecutionCount, 1)
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift:41]: I'd suggest making these assertions async since the notification callback doesn't have to happen synchronously. `AssertAsync.willBeEqual` here ...

File: StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+        // Release the observer
+        observer = nil
+
+        // Send event notification
+        notificationCenter.post(eventNotification)
+        // Assert callback is not called one more time
+        XCTAssertEqual(callbackExecutionCount, 1)
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift:49]: ... and `AssertAsync.staysEqual` here

File: StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                                     tranform: { $0 as? HealthCheckEvent },
+                                     callback: { receivedEvent = $0 })
+        _ = observer
+
+        // Send event notification
+        notificationCenter.post(eventNotification)
+
+        // Assert event is received
+        XCTAssertNotNil(receivedEvent)
+        XCTAssertEqual(receivedEvent?.connectionId, eventToDeliver.connectionId)
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift:66]: Same here

File: StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    func test_callbackIsCalled_ifEventCastFails() {
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift:69]: callback is **not** called

File: StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        let observer = EventObserver(notificationCenter: notificationCenter,
+                                     tranform: { $0 as? MemberAddedEvent },
+                                     callback: { receivedEvent = $0 })
+        _ = observer
+
+        // Send event notification
+        notificationCenter.post(eventNotification)
+
+        // Assert none event is received
+        XCTAssertNil(receivedEvent)
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift:82]: `AssertAsync.staysTrue(receivedEvent == nil)`

File: StreamChatClient_v3/Workers/EventObservers/MemberEventObserver_Tests.swift [hunk 8]
```diff
@@ -0,0 +1,126 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+@testable import StreamChatClient_v3
+import XCTest
+
+final class MemberEventObserver_Tests: XCTestCase {
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Workers/EventObservers/MemberEventObserver_Tests.swift:8]: Do you think we need such detailed tests for this? It duplicates a lot the tests of `EventObserver`. Maybe it would be enough to just test the `filter` is set up correctly?

File: StreamChatClient_v3/Controllers/ChannelController.swift [hunk 9]
```diff
@@ -119,6 +119,7 @@ public class ChannelControllerGeneric<ExtraData: ExtraDataTypes>: Controller, De
         return observer
     }()
     
+    private var eventObservers = [EventObserver]()
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Controllers/ChannelController.swift:122]: nit: we usually use the following syntax for this:
```swift
private var eventObservers: [EventObserver] = []
```

File: StreamChatClient_v3/Controllers/ChannelController_Tests.swift [hunk 10]
```diff
@@ -223,10 +223,18 @@ class ChannelController_Tests: StressTestCase {
         let channel: Channel = client.databaseContainer.viewContext.loadChannel(cid: channelId)!
         assert(channel.latestMessages.count == 1)
         let message: Message = channel.latestMessages.first!
+
+        // Send notification with event happened in the observed channel
+        let event = try MemberAddedEvent.dummy(cid: controller.channelQuery.cid, userId: .unique)
+        let notification = Notification(newEventReceived: event, sender: self)
+        client.webSocketClient.notificationCenter.post(notification)
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Controllers/ChannelController_Tests.swift:230]: Let's create a separate test method for this, please. We will have many delegate methods, we need to start structuring it.

File: StreamChatClient_v3/Controllers/ChannelController_Tests.swift [hunk 11]
```diff
@@ -569,3 +585,17 @@ private class TestDelegateGeneric: QueueAwareDelegate, ChannelControllerDelegate
...
+                                                          role: .member,
+                                                          createdAt: .unique,
+                                                          updatedAt: .unique)
+
+        let payload = EventPayload<DefaultDataTypes>(eventType: .memberAdded,
+                                                     cid: cid,
+                                                     memberContainer: .init(member: member))
+        return try .init(from: payload)
+    }
+}
```
Comment (VojtaStavik) [inline StreamChatClient_v3/Controllers/ChannelController_Tests.swift]: You can just create `TestMemberEvent: MemberEvent`, no need to use the real event types here.

File: StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift (hunk 2 above)
Comment (evsaev) [reply StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift]: fixed in c79a9c3ebd4f36f3dbb1bcc63f4b4a544765467a

File: StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift (hunk 3 above)
Comment (evsaev) [reply StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift]: fixed in c79a9c3ebd4f36f3dbb1bcc63f4b4a544765467a

File: StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift (hunk 4 above)
Comment (evsaev) [reply StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift]: fixed in c79a9c3ebd4f36f3dbb1bcc63f4b4a544765467a

File: StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift (hunk 5 above)
Comment (evsaev) [reply StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift]: fixed in c79a9c3ebd4f36f3dbb1bcc63f4b4a544765467a

File: StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift (hunk 6 above)
Comment (evsaev) [reply StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift]: fixed in c79a9c3ebd4f36f3dbb1bcc63f4b4a544765467a

File: StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift (hunk 7 above)
Comment (evsaev) [reply StreamChatClient_v3/Workers/EventObservers/EventObserver_Tests.swift]: fixed in c79a9c3ebd4f36f3dbb1bcc63f4b4a544765467a

File: StreamChatClient_v3/Workers/EventObservers/MemberEventObserver_Tests.swift (hunk 8 above)
Comment (evsaev) [reply StreamChatClient_v3/Workers/EventObservers/MemberEventObserver_Tests.swift]: fixed in eaa477096e31097b6278c929970beb799580a3ae

File: StreamChatClient_v3/Controllers/ChannelController.swift (hunk 9 above)
Comment (evsaev) [reply StreamChatClient_v3/Controllers/ChannelController.swift]: fixed in 791b4a3673a2127949ad673e06f1ae90366330f9

File: StreamChatClient_v3/Controllers/ChannelController_Tests.swift (hunk 10 above)
Comment (evsaev) [reply StreamChatClient_v3/Controllers/ChannelController_Tests.swift]: fixed in 791b4a3673a2127949ad673e06f1ae90366330f9

File: StreamChatClient_v3/Controllers/ChannelController_Tests.swift (hunk 11 above)
Comment (evsaev) [reply StreamChatClient_v3/Controllers/ChannelController_Tests.swift]: fixed in 791b4a3673a2127949ad673e06f1ae90366330f9

File: StreamChatClient_v3/Workers/EventObservers/MemberEventObserver_Tests.swift (hunk 8 above)
Comment (VojtaStavik) [reply StreamChatClient_v3/Workers/EventObservers/MemberEventObserver_Tests.swift]: We either didn't understand each other or this is not fixed :)

File: Sources_v3/Controllers/ChannelController_Tests.swift [hunk 12]
```diff
@@ -216,6 +216,52 @@ class ChannelController_Tests: StressTestCase {
...
+        // Send notification with event happened in the observed channel
+        let event = TestMemberEvent(cid: controller.channelQuery.cid, userId: .unique)
+        let notification = Notification(newEventReceived: event, sender: self)
+        client.webSocketClient.notificationCenter.post(notification)
+
+        // Assert the event is received
+        AssertAsync {
+            Assert.willBeTrue(delegate.didReceiveMemberEvent_event is TestMemberEvent)
+            Assert.willBeEqual(delegate.didReceiveMemberEvent_event?.userId, event.userId)
+            Assert.willBeEqual(delegate.didReceiveMemberEvent_event?.cid, event.cid)
```
Comment (VojtaStavik) [inline Sources_v3/Controllers/ChannelController_Tests.swift:239]: You can make `TestMemberEvent` equatable and just assert it's the correct one.

File: StreamChatClient_v3/Workers/EventObservers/MemberEventObserver_Tests.swift (hunk 8 above)
Comment (evsaev) [reply StreamChatClient_v3/Workers/EventObservers/MemberEventObserver_Tests.swift]: Hm, you are right 🤔  Did these changes but lost somewhere on the go. Finally fixed in b40eab6d84232560c2058e5f5c22e7f0d89ed128

File: Sources_v3/Controllers/ChannelController_Tests.swift (hunk 12 above)
Comment (evsaev) [reply Sources_v3/Controllers/ChannelController_Tests.swift]: fixed in e89ecd4e32c9555e1079d9112b1168c2467ed4f4

File: Sources_v3/Workers/EventObservers/EventObserver_Tests.swift [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+    override func setUp() {
+        super.setUp()
+
+        notificationCenter = NotificationCenter()
+        eventToDeliver = HealthCheckEvent(connectionId: .unique)
+    }
+
+    override func tearDown() {
+        super.tearDown()
```
Comment (VojtaStavik) [inline Sources_v3/Workers/EventObservers/EventObserver_Tests.swift:26]: nit: `super.tearDown()` should be called at the end of this function. You want to clean up your things first before the superclass cleans its things. In this very case, it doesn't really matter, since it's not mandatory to call super when subclassing from `XCTestCase`. However, sometimes we use a different superclass (like `StressTestCase`) and it can matter there. It's just a best practice we established internally to save us from headaches in the future.

File: Sources_v3/Workers/EventObservers/MemberEventObserver.swift [hunk 14]
```diff
@@ -0,0 +1,32 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+final class MemberEventObserver: EventObserver {
```
Comment (VojtaStavik) [inline Sources_v3/Workers/EventObservers/MemberEventObserver.swift:7]: nit: Please add a simple description of what this class does

File: Sources_v3/Workers/EventObservers/MemberEventObserver_Tests.swift [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+    // MARK: - Setup
+
+    override func setUp() {
+        super.setUp()
+
+        eventNotificationCenter = NotificationCenter()
+    }
+
+    override func tearDown() {
+        super.tearDown()
```
Comment (VojtaStavik) [inline Sources_v3/Workers/EventObservers/MemberEventObserver_Tests.swift:21]: see above

File: Sources_v3/Workers/EventObservers/MemberEventObserver.swift (hunk 14 above)
Comment (VojtaStavik) [reply Sources_v3/Workers/EventObservers/MemberEventObserver.swift:7]: @evsaev You ignored my comment :)
